cogs/birthday/birthday.py

```diff
@@ -510,56 +557,110 @@ async def setSelfBirthday(self, ctx: Context, *, birthday: MonthDayConverter):
+                self.logger.info(
+                    "%s#%s (%s) added their birthday as %s",
+                    ctx.author.name,
+                    ctx.author.discriminator,
+                    ctx.author.id,
+                    birthdayStr,
+                )
 
-            await birthdayConfig.get_attr(KEY_BDAY_MONTH).set(birthday.month)
-            await birthdayConfig.get_attr(KEY_BDAY_DAY).set(birthday.day)
-            await birthdayConfig.get_attr(KEY_ADDED_BEFORE).set(True)
+                # explicitly check to see if the role should be applied to the user
+                # if the month and day just so happen to be the same as it is now.
+                await self.checkBirthday()
+                return
 
-            await ctx.author.send(
-                f"{headerGood}: Successfully set your birthday to "
-                f"{spoiler(bold(birthdayStr, escape_formatting=False), escape_formatting=False)}."
+            noDmStr = "\n".join(
+                (
+                    "You have disabled DMs from this server. Would you "
+                    "still like to continue here? All messages containing your "
+                    "birthday will be deleted after a short delay.",
+                    f"Type {bold('`yes`', escape_formatting=False)} to confirm.",
+                )
             )
+            try:
+                await mainFlow(ctx.author.dm_channel, carefree=True)
+                return
+            except discord.Forbidden:
+                await ctx.send(noDmStr)
 
-            self.logger.info(
-                "%s#%s (%s) added their birthday as %s",
-                ctx.author.name,
-                ctx.author.discriminator,
-                ctx.author.id,
-                birthdayStr,
-            )
+                def check(msg: discord.Message):
+                    return msg.author == ctx.author and msg.channel == ctx.channel
 
-            # explicitly check to see if user should be added to role, if the month
-            # and day just so happen to be the same as it is now.
-            await self.checkBirthday()
-            return
+                try:
+                    response = await self.bot.wait_for(
+                        "message", timeout=10.0, check=check
+                    )
+                except asyncio.TimeoutError:
+                    await ctx.send(
+                        f"{headerBad}: You took too long. Not setting your birthday."
+                    )
+                    return
 
-        raise Exception("Error while accessing member's birthday config. This should not happen!")
+                if response.content.lower() != "yes":
+                    await ctx.send(f"{headerBad}: Declined. Not setting your birthday.")
+                    return
+                await mainFlow(ctx.channel, carefree=False)
+                return
+
+        raise Exception(
+            "Error while accessing member's birthday config. This should not happen!"
+        )
```

Comment:
Somewhat out of this PR's scope but it's related to the `mainFlow` function which is the main flow of this method `setSelfBirthday` so I'm having this comment while we're here.

Quick refactor: we can just remove this `raise`.

```suggestion
                if response.content.lower() != "yes":
                    await ctx.send(f"{headerBad}: Declined. Not setting your birthday.")
                    return

                await mainFlow(ctx.channel, carefree=False)
```

This exception can be dropped because our `Config` instance has `force_registration=True` so an `AttributeError` will be raised when we fail to access member's birthday config. Also, in case this exception is needed, then it should stay closer to the attribute accessing calls (i.e., should stay inside `mainFlow`) and we should change it to `RuntimeError` because `Exception` is supposed to be a base class for others to derive.

I see that it's redundant so my suggestion is to just drop it.

Reply:
Removed